src/main/java/graphql/util/RecursionState.java

```diff
@@ -1,70 +1,97 @@
 package graphql.util;
 
+import graphql.Assert;
 import graphql.Internal;
 
 import java.util.ArrayDeque;
 import java.util.Collection;
 import java.util.Deque;
-import java.util.HashMap;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
-import java.util.Optional;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
-import java.util.stream.Collectors;
 
 import static graphql.Assert.assertNotNull;
 
 @Internal
-abstract class RecursionState<T> {
+public class RecursionState<T> {
 
-    private final Deque<TraverserContext<?>> delegate;
-    private final Map<T, Object> visitedMap = new ConcurrentHashMap<>();
+    private Object initialData;
 
-    public RecursionState() {
-        this(new ArrayDeque<>(32));
+    public enum Type {
+        STACK,
+        QUEUE
     }
 
-    public RecursionState(Deque<? super TraverserContext<T>> delegate) {
-        this.delegate = (Deque<TraverserContext<?>>) assertNotNull(delegate);
+    private final Deque<Object> state;
+    private final Set<T> visited = new LinkedHashSet<>();
+    private final Type type;
+
+
+    public enum Marker {
+        END_LIST
     }
 
-    public TraverserContext<T> peek() {
-        return (TraverserContext<T>) delegate.peek();
+    public RecursionState(Type type, Object initialData) {
+        this.initialData = initialData;
+        this.state = new ArrayDeque<>(32);
+        this.type = assertNotNull(type);
     }
 
-    public abstract TraverserContext<T> pop();
 
-    public abstract void pushAll(TraverserContext<T> o, Function<? super T, ? extends List<T>> getChildren);
+    public Object peek() {
+        return state.peek();
+    }
+
+
+    public Object pop() {
+        return this.state.pop();
+    }
+
+    public void pushAll(TraverserContext<T> o, Function<? super T, ? extends List<T>> getChildren) {
+        if (type == Type.QUEUE) {
```

Comment:
In my opinion, this runtime check adds complexity to this algorithm (additional O(N), where N is the number of Nodes in the tree).

Any particular reason why sub-typing with compile time binding is not used?

Reply:
because having two separate classes which are just different by a slightly different `pushAll` methods are more complicated than having this check here. (Both classes implemented the `pop` method basically the same way, but with a different method on `Deque. This became immediately clear after I merged them, but before that it was not obvious)

But having two inline classes in RecursionState might be a good alternative.

Reply:
I vote for 2 inline classes and get rid of the extra check for the queue type

Reply:
ok, changed it.